server/migrator/Schema.java

```diff
@@ -0,0 +1,181 @@
+package grakn.core.server.migrator;
+
+import grakn.core.Grakn;
+import grakn.core.common.exception.GraknException;
+import grakn.core.common.parameters.Arguments;
+import grakn.core.concept.type.AttributeType;
+import grakn.core.concept.type.EntityType;
+import grakn.core.concept.type.RelationType;
+import grakn.core.concept.type.ThingType;
+import grakn.core.logic.Rule;
+import graql.lang.pattern.Conjunction;
+import graql.lang.pattern.Pattern;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static grakn.common.collection.Collections.list;
+import static grakn.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+
+public class Schema {
+
+    private static String SEMICOLON_NEWLINE_X2 = ";\n\n";
+    private static String COMMA_NEWLINE_INDENT = ",\n" + indent(1);
+    private final Grakn grakn;
+    private final String database;
+
+    public Schema(final Grakn grakn, final String database) {
+        this.grakn = grakn;
+        this.database = database;
+    }
+
+    public String getSchema() {
+        StringBuilder builder = new StringBuilder();
+        builder.append("define\n\n");
+        try (final Grakn.Session session = grakn.session(database, Arguments.Session.Type.DATA);
+             final Grakn.Transaction tx = session.transaction(Arguments.Transaction.Type.READ)) {
+            tx.concepts()
+                    .getRootAttributeType()
+                    .getSubtypes()
+                    .filter(x -> !x.isRoot())
+                    .sorted(Comparator.comparing(x -> x.getLabel().name()))
+                    .forEach(x -> writeAttributeType(builder, x));
+            tx.concepts()
+                    .getRootRelationType()
+                    .getSubtypes()
+                    .filter(x -> !x.isRoot())
+                    .sorted(Comparator.comparing(x -> x.getLabel().name()))
+                    .forEach(x -> writeRelationType(builder, x));
+            tx.concepts()
+                    .getRootEntityType()
+                    .getSubtypes()
+                    .filter(x -> !x.isRoot())
+                    .sorted(Comparator.comparing(x -> x.getLabel().name()))
+                    .forEach(x -> writeEntityType(builder, x));
```

Comment:
Let's extract `.getSubtypes().filter(x -> !x.isRoot()).sorted(Comparator.comparing(x -> x.getLabel().name()))` into its own method, something like:
```java
private <TYPE extends Type> Stream<TYPE> subtypesSortedByLabel(TYPE type) {
    return type.getSubtypes().filter(x -> !x.isRoot()).sorted(Comparator.comparing(x -> x.getLabel().name()));
}
```

Reply:
discussed in person, this is harder to do.